sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -0,0 +1,179 @@
+        {
+            length = _bytes.Length;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellation)
+        {
+            byte[] buffer = _bytes.ToArray();
+            stream.Write(buffer, 0, buffer.Length);
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+            => await stream.WriteAsync(_bytes, cancellation).ConfigureAwait(false);
+
+        public override void Dispose() { }
+    }
+
+    private sealed class ModelBinaryContent<T> : BinaryContent where T : IPersistableModel<T>
+    {
+        private readonly T _model;
+        private readonly ModelReaderWriterOptions _options;
+
+        // Used when _model is an IJsonModel
+        private ModelWriter? _writer;
+
+        // Used when _model is an IModel
+        private BinaryData? _data;
+
+        public ModelBinaryContent(T model, ModelReaderWriterOptions options)
+        {
+            _model = model;
+            _options = options;
+        }
+
+        private ModelWriter Writer
+        {
+            get
+            {
+                if (_model is not IJsonModel<T> jsonModel)
+                {
+                    throw new InvalidOperationException("Cannot use Writer with non-IJsonModel model type.");
+                }
+
+                _writer ??= new ModelWriter((IJsonModel<object>)jsonModel, _options);
+                return _writer;
+            }
+        }
+
+        private BinaryData Data
+        {
+            get
+            {
+                if (_model is IJsonModel<T> && _options.Format == "J")
+                {
+                    throw new InvalidOperationException("Should use ModelWriter instead of _model.Write with IJsonModel.");
+                }
+
+                _data ??= _model.Write(_options);
+                return _data;
+            }
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            if (_model is IJsonModel<T> && _options.Format == "J")
+            {
+                return Writer.TryComputeLength(out length);
+            }
+
+            length = Data.ToMemory().Length;
+            return true;
+        }
+
+#if NETFRAMEWORK || NETSTANDARD2_0
+        private byte[]? _bytes;
+        private byte[] Bytes => _bytes ??= Data.ToArray();
+#endif
+
+        public override void WriteTo(Stream stream, CancellationToken cancellation)
+        {
+            if (_model is IJsonModel<T> && _options.Format == "J")
+            {
+                Writer.CopyTo(stream, cancellation);
+                return;
+            }
+
+#if NETFRAMEWORK || NETSTANDARD2_0
+            stream.Write(Bytes, 0, Bytes.Length);
+#else
+            stream.Write(Data.ToMemory().Span);
+#endif
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+        {
+            if (_model is IJsonModel<T> && _options.Format == "J")
```

Comment:
I think we should perhaps expose [this functionality](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs#L34) in an internal static method on MRW so that we can duplicate the logic.  It will have a higher hit rate on the faster API which will be a good thing.

Reply:
Per offline discussion, will address in a separate PR, and thanks for raising, @m-nash!